File: tests/src/com.google.protobuf/protobuf-javalite/3.25.5/src/test/java/com_google_protobuf/protobuf_javalite/MessageSchemaTest.java
/*
 * Copyright and related rights waived via CC0
 *
 * You should have received a copy of the CC0 legalcode along with this
 * work. If not, see <http://creativecommons.org/publicdomain/zero/1.0/>.
 */
package com_google_protobuf.protobuf_javalite;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertThrows;

import com.google.protobuf.GeneratedMessageLite;
import com.google.protobuf.InvalidProtocolBufferException;
import com.google.protobuf.Parser;
import org.junit.jupiter.api.Test;

public class MessageSchemaTest {
    @Test
    public void parserBuildsSchemaForDeclaredFieldBackedMessage() throws InvalidProtocolBufferException {
        DeclaredFieldMessage message = DeclaredFieldMessage.parseFrom(new byte[] {8, 42});

        assertEquals(42, message.getValue());
    }

    @Test
    public void parserReportsMissingMessageInfoFieldAfterScanningDeclaredFields() {
        assertThrows(RuntimeException.class, () -> MissingFieldMessage.parseFrom(new byte[0]));
    }

    public static final class DeclaredFieldMessage
            extends GeneratedMessageLite<DeclaredFieldMessage, DeclaredFieldMessageBuilder> {
        private static final DeclaredFieldMessage DEFAULT_INSTANCE = new DeclaredFieldMessage();
        private static volatile Parser<DeclaredFieldMessage> parser;

        static {
            registerDefaultInstance(DeclaredFieldMessage.class, DEFAULT_INSTANCE);
        }

        private int value_;

        private DeclaredFieldMessage() {
        }

        public int getValue() {
            return value_;
        }

        public static DeclaredFieldMessage parseFrom(byte[] data) throws InvalidProtocolBufferException {
            return GeneratedMessageLite.parseFrom(DEFAULT_INSTANCE, data);
        }

        @Override
        protected Object dynamicMethod(MethodToInvoke method, Object arg0, Object arg1) {
            switch (method) {
                case NEW_MUTABLE_INSTANCE:
                    return new DeclaredFieldMessage();
                case NEW_BUILDER:
                    return new DeclaredFieldMessageBuilder();
                case BUILD_MESSAGE_INFO:
                    return newMessageInfo(
                            DEFAULT_INSTANCE,
                            "\u0000\u0001\u0000\u0000\u0001\u0001\u0001\u0000\u0000\u0000\u0001\u0004",
                            new Object[] {"value_"});
                case GET_DEFAULT_INSTANCE:
                    return DEFAULT_INSTANCE;
                case GET_PARSER:
                    Parser<DeclaredFieldMessage> result = parser;
                    if (result == null) {
                        synchronized (DeclaredFieldMessage.class) {
                            result = parser;
                            if (result == null) {
                                result = new DefaultInstanceBasedParser<>(DEFAULT_INSTANCE);
                                parser = result;
                            }
                        }
                    }
                    return result;
                case GET_MEMOIZED_IS_INITIALIZED:
                    return (byte) 1;
                case SET_MEMOIZED_IS_INITIALIZED:
                    return null;
                default:
                    throw new UnsupportedOperationException();
            }
        }
    }

    public static final class DeclaredFieldMessageBuilder
            extends GeneratedMessageLite.Builder<DeclaredFieldMessage, DeclaredFieldMessageBuilder> {
        private DeclaredFieldMessageBuilder() {
            super(DeclaredFieldMessage.DEFAULT_INSTANCE);
        }
    }

    public static final class MissingFieldMessage
            extends GeneratedMessageLite<MissingFieldMessage, MissingFieldMessageBuilder> {
        private static final MissingFieldMessage DEFAULT_INSTANCE = new MissingFieldMessage();
        private static volatile Parser<MissingFieldMessage> parser;

        static {
            registerDefaultInstance(MissingFieldMessage.class, DEFAULT_INSTANCE);
        }

        private MissingFieldMessage() {
        }

        public static MissingFieldMessage parseFrom(byte[] data) throws InvalidProtocolBufferException {
            return GeneratedMessageLite.parseFrom(DEFAULT_INSTANCE, data);
        }

        @Override
        protected Object dynamicMethod(MethodToInvoke method, Object arg0, Object arg1) {
            switch (method) {
                case NEW_MUTABLE_INSTANCE:
                    return new MissingFieldMessage();
                case NEW_BUILDER:
                    return new MissingFieldMessageBuilder();
                case BUILD_MESSAGE_INFO:
                    return newMessageInfo(
                            DEFAULT_INSTANCE,
                            "\u0000\u0001\u0000\u0000\u0001\u0001\u0001\u0000\u0000\u0000\u0001\u0004",
                            new Object[] {"missing_"});
                case GET_DEFAULT_INSTANCE:
                    return DEFAULT_INSTANCE;
                case GET_PARSER:
                    Parser<MissingFieldMessage> result = parser;
                    if (result == null) {
                        synchronized (MissingFieldMessage.class) {
                            result = parser;
                            if (result == null) {
                                result = new DefaultInstanceBasedParser<>(DEFAULT_INSTANCE);
                                parser = result;
                            }
                        }
                    }
                    return result;
                case GET_MEMOIZED_IS_INITIALIZED:
                    return (byte) 1;
                case SET_MEMOIZED_IS_INITIALIZED:
                    return null;
                default:
                    throw new UnsupportedOperationException();
            }
        }
    }

    public static final class MissingFieldMessageBuilder
            extends GeneratedMessageLite.Builder<MissingFieldMessage, MissingFieldMessageBuilder> {
        private MissingFieldMessageBuilder() {
            super(MissingFieldMessage.DEFAULT_INSTANCE);
        }
    }
}

File: tests/src/org.apache.pekko/pekko-protobuf-v3_3/1.0.1/src/test/scala/org_apache_pekko/pekko_protobuf_v3_3/MessageSchemaTest.scala
/*
 * Copyright and related rights waived via CC0
 *
 * You should have received a copy of the CC0 legalcode along with this
 * work. If not, see <http://creativecommons.org/publicdomain/zero/1.0/>.
 */
package org_apache_pekko.pekko_protobuf_v3_3

import org.apache.pekko.protobufv3.internal.GeneratedMessageLite
import org.assertj.core.api.Assertions.assertThat
import org.assertj.core.api.Assertions.catchThrowable
import org.assertj.core.api.ThrowableAssert.ThrowingCallable
import org.junit.jupiter.api.Test

class MessageSchemaTest {
  @Test
  def createsLiteSchemaFromGeneratedMessageInfoFields(): Unit = {
    val message: MessageSchemaReflectableMessage = new MessageSchemaReflectableMessage()

    message.setText("schema field lookup")

    assertThat(message.toByteArray).isNotEmpty()
    assertThat(message.hashCode()).isEqualTo(message.hashCode())
  }

  @Test
  def reportsGeneratedMessageInfoFieldsMissingFromMessageClass(): Unit = {
    val message: MessageSchemaMissingFieldMessage = new MessageSchemaMissingFieldMessage()

    val thrown: Throwable = catchThrowable(new ThrowingCallable {
      override def call(): Unit = message.hashCode()
    })

    assertThat(thrown).isInstanceOf(classOf[RuntimeException])
    assertThat(thrown.getMessage).contains("Field absent_")
    assertThat(thrown.getMessage).contains("not found")
  }
}

final class MessageSchemaReflectableMessage
  extends GeneratedMessageLite[MessageSchemaReflectableMessage, MessageSchemaReflectableBuilder] {
  private var `text_`: String = ""

  def setText(value: String): Unit = {
    `text_` = value
  }

  override protected def dynamicMethod(
    methodToInvoke: GeneratedMessageLite.MethodToInvoke,
    argument0: Object,
    argument1: Object
  ): Object = {
    methodToInvoke match {
      case GeneratedMessageLite.MethodToInvoke.NEW_MUTABLE_INSTANCE =>
        new MessageSchemaReflectableMessage()
      case GeneratedMessageLite.MethodToInvoke.NEW_BUILDER =>
        new MessageSchemaReflectableBuilder()
      case GeneratedMessageLite.MethodToInvoke.BUILD_MESSAGE_INFO =>
        GeneratedMessageLiteTest.newMessageInfo(
          MessageSchemaReflectableMessage.defaultInstance,
          MessageSchemaMessageInfo.SingleUtf8StringField,
          Array[AnyRef]("text_")
        )
      case GeneratedMessageLite.MethodToInvoke.GET_DEFAULT_INSTANCE =>
        MessageSchemaReflectableMessage.defaultInstance
      case GeneratedMessageLite.MethodToInvoke.GET_PARSER =>
        new GeneratedMessageLite.DefaultInstanceBasedParser[MessageSchemaReflectableMessage](
          MessageSchemaReflectableMessage.defaultInstance
        )
      case GeneratedMessageLite.MethodToInvoke.GET_MEMOIZED_IS_INITIALIZED =>
        java.lang.Byte.valueOf(1.toByte)
      case GeneratedMessageLite.MethodToInvoke.SET_MEMOIZED_IS_INITIALIZED =>
        null
    }
  }
}

object MessageSchemaReflectableMessage {
  val defaultInstance: MessageSchemaReflectableMessage = new MessageSchemaReflectableMessage()
}

final class MessageSchemaReflectableBuilder
  extends GeneratedMessageLite.Builder[MessageSchemaReflectableMessage, MessageSchemaReflectableBuilder](
    MessageSchemaReflectableMessage.defaultInstance
  )

final class MessageSchemaMissingFieldMessage
  extends GeneratedMessageLite[MessageSchemaMissingFieldMessage, MessageSchemaMissingFieldBuilder] {

  override protected def dynamicMethod(
    methodToInvoke: GeneratedMessageLite.MethodToInvoke,
    argument0: Object,
    argument1: Object
  ): Object = {
    methodToInvoke match {
      case GeneratedMessageLite.MethodToInvoke.NEW_MUTABLE_INSTANCE =>
        new MessageSchemaMissingFieldMessage()
      case GeneratedMessageLite.MethodToInvoke.NEW_BUILDER =>
        new MessageSchemaMissingFieldBuilder()
      case GeneratedMessageLite.MethodToInvoke.BUILD_MESSAGE_INFO =>
        GeneratedMessageLiteTest.newMessageInfo(
          MessageSchemaMissingFieldMessage.defaultInstance,
          MessageSchemaMessageInfo.SingleUtf8StringField,
          Array[AnyRef]("absent_")
        )
      case GeneratedMessageLite.MethodToInvoke.GET_DEFAULT_INSTANCE =>
        MessageSchemaMissingFieldMessage.defaultInstance
      case GeneratedMessageLite.MethodToInvoke.GET_PARSER =>
        new GeneratedMessageLite.DefaultInstanceBasedParser[MessageSchemaMissingFieldMessage](
          MessageSchemaMissingFieldMessage.defaultInstance
        )
      case GeneratedMessageLite.MethodToInvoke.GET_MEMOIZED_IS_INITIALIZED =>
        java.lang.Byte.valueOf(1.toByte)
      case GeneratedMessageLite.MethodToInvoke.SET_MEMOIZED_IS_INITIALIZED =>
        null
    }
  }
}

object MessageSchemaMissingFieldMessage {
  val defaultInstance: MessageSchemaMissingFieldMessage = new MessageSchemaMissingFieldMessage()
}

final class MessageSchemaMissingFieldBuilder
  extends GeneratedMessageLite.Builder[MessageSchemaMissingFieldMessage, MessageSchemaMissingFieldBuilder](
    MessageSchemaMissingFieldMessage.defaultInstance
  )

object MessageSchemaMessageInfo {
  val SingleUtf8StringField: String = "\u0000\u0001\u0000\u0000\u0001\u0001\u0001\u0000\u0000\u0000\u0001\u0208"
}
